### DISCUSSION TITLE
Give user the ability to "cross-post" to Hugging Face

### DISCUSSION THREAD

**Discussion by JustMaier** [civitai/civitai#67] [Ideas]
The thought on this one is that we would make it easier for creators to post to both services at once. They'd just need to put in their Hugging Face Token and then we'd be able to do the following on their behalf using the [Hugging Face Hub API](https://huggingface.co/docs/hub/api):
1. Create a Repo with a good looking README based on the examples etc that they provided
1. Upload the checkpoint to the Repo
1. Run [SD to Diffusers](https://huggingface.co/spaces/anzorq/sd-to-diffusers) to enable them for Diffusers. _Merge the PR_ ([Documented here](https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L3650)).
1. Run [SD Space Creator](https://huggingface.co/spaces/anzorq/sd-space-creator) to make a Hugging Face space for them
1. Add an entry to the Run Now listing for the model that points to the Hugging Face SD Space for the model.

To get this working I believe we'd have to do a few things:
- [ ] Add a toggle to the Model editor
- [ ] Add API endpoints to the HF spaces
    - [ ] Determine if there is a webhook system or if we need to have some sort of long-running background job. Build accordingly.
- [ ] Ask HF to add an endpoint for Merging PRs
- [ ] Add jobs for requesting the run of the HF Spaces
- [ ] Add a job that adds the HF Space as a Run Now listing (Definition of a Run Now listing is still pending our development)

**Comment by osanseviero** [comment]
Hi! I'm Omar from Hugging Face. We already provide a mechanism to programmatically merge Pull Requests! See https://github.com/huggingface/huggingface_hub/blob/main/src/huggingface_hub/hf_api.py#L3650

**Reply by JustMaier** [reply]
Awesome! I didn't see it in the [Hub API Endpoints docs](https://huggingface.co/docs/hub/api). This is great! At this point, I think we just need to build it

**Reply by osanseviero** [reply]
Thanks, we will improve that API page :)

**Comment by keturn** [comment]
InvokeAI's upcoming release loads diffusers-format models natively.

While we do have a workflow that can load a CompVis-format file, we've found that some models (especially merges) have data that's been put together in unexpected ways, and the conversion process is not always seamless. I think we'll have better results and a more efficient loading process if model trainers are able to provide an official diffusers-format distribution.

If trainers take advantage of 🤗's generous offer to host models and cross-post to there, we need only the 🤗 model ID. It would be sufficient to display that on the model page and include it in the API.

If Civitai wants to offer their own hosting for diffusers-format models, then we'll want to figure out a more detailed API.

I had full model weights in mind when writing that, but

- for Textual Inversion, Hugging Face has a [Concepts Library](https://huggingface.co/sd-concepts-library) we integrate with. Cross-posting to there is probably a slightly different workflow than for full models.
- as of [diffusers 0.12](https://github.com/huggingface/diffusers/releases/tag/v0.12.0), the library now supports training and loading LoRA weights. InvokeAI doesn't have integration for that _yet_, but it's a highly requested feature we'll want to add soon. I don't know enough about LoRA yet to know if there are any conversion tricks or metadata files necessary for those.
